### PR TITLE
[Frost DK] 8.3 APL Updates

### PR DESCRIPTION
**Pull request by nyterage**
some general improvements for icecap related to rune use and essence timing.

rewrote Frostwyrms Fury to be a bit more readable, and improved it slightly.

added some labeling for the AoE rotation, doesn't effect dps, but its easier to find now. 

Changes do result in a dps loss with some azerite traits, but a major increase with others. after much discussion with tzo, bicep and chew, i got the go ahead from them for these changes, as their increase for 8.3 is quite great. sim below contains both 8.2.5 sims, as well as 8.3 sims. 

https://www.raidbots.com/simbot/report/dhFgdqM4hYE1TYR7WveZY3

### CODE REVIEW

File: engine/class_modules/sc_death_knight.cpp [hunk 1]
```diff
@@ -7782,15 +7783,18 @@ void death_knight_t::default_apl_frost()
...
   standard -> add_action( this, "Frost Strike", "if=runic_power.deficit<(15+talent.runic_attenuation.enabled*3)" );
   standard -> add_talent( this, "Frostscythe", "if=buff.killing_machine.up&rune.time_to_4>=gcd" );
   standard -> add_action( this, "Obliterate", "if=runic_power.deficit>(25+talent.runic_attenuation.enabled*3)" );
   standard -> add_action( this, "Frost Strike" );
   standard -> add_talent( this, "Horn of Winter" );
   standard -> add_action( "arcane_torrent" );
-
-  aoe -> add_action( this, "Remorseless Winter", "if=talent.gathering_storm.enabled|(azerite.frozen_tempest.rank&spell_targets.remorseless_winter>=3&!buff.rime.up)" );
+	
+  // AoE Rotation
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: What's the point of this?
The additional "AoE Rotation" parameter outputs the "#AoE Rotation" comment in the default profiles.
The variable's name for this specific APL is already self explanatory so I don't think it warrants a c++ side code comment

File: engine/class_modules/sc_death_knight.cpp [hunk 2]
```diff
@@ -7782,15 +7783,18 @@ void death_knight_t::default_apl_frost()
   standard -> add_action( this, "Remorseless Winter", "", "Standard single-target rotation" );
   standard -> add_action( this, "Frost Strike", "if=cooldown.remorseless_winter.remains<=2*gcd&talent.gathering_storm.enabled" );
   standard -> add_action( this, "Howling Blast", "if=buff.rime.up" );
+  standard -> add_action( this, "Frostscythe", "if=talent.icecap.enabled&buff.killing_machine.up" );
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: Please use buff.killing_machine.react to account for the reaction time after a proc.

File: engine/class_modules/sc_death_knight.cpp [hunk 3]
```diff
@@ -7710,15 +7710,16 @@ void death_knight_t::default_apl_frost()
...
   cooldowns -> add_action( this, "Empower Rune Weapon", "if=(cooldown.pillar_of_frost.ready|target.1.time_to_die<20)&talent.breath_of_sindragosa.enabled&runic_power>60" );
-
+  cooldowns -> add_action( this, "Empower Rune Weapon", "if=talent.icecap.enabled" );
+	
   // Cold Heart and Frostwyrm's Fury
   cooldowns -> add_action( "call_action_list,name=cold_heart,if=talent.cold_heart.enabled&((buff.cold_heart.stack>=10&debuff.razorice.stack=5)|target.1.time_to_die<=gcd)" );
-  cooldowns -> add_talent( this, "Frostwyrm's Fury", "if=(buff.pillar_of_frost.remains<=gcd|(buff.pillar_of_frost.remains<8&buff.unholy_strength.remains<=gcd&buff.unholy_strength.up))&buff.pillar_of_frost.up&azerite.icy_citadel.rank<=1" );
-  cooldowns -> add_talent( this, "Frostwyrm's Fury", "if=(buff.icy_citadel.remains<=gcd|(buff.icy_citadel.remains<8&buff.unholy_strength.remains<=gcd&buff.unholy_strength.up))&buff.icy_citadel.up&azerite.icy_citadel.rank>=2" );
+  cooldowns -> add_talent( this, "Frostwyrm's Fury", "if=(azerite.icy_citadel.rank<=1&buff.pillar_of_frost.up&(buff.pillar_of_frost.remains<=gcd|buff.unholy_strength.remains<=gcd|buff.icy_citadel.remains<=gcd&buff.icy_citadel.up))" );
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: These two lines have the same conditions, making the behavior the same regardless of Icy Citadel ranks. Is that intended? Moreover, even though you check for buff.unholy_strength.remains<=gcd, you don't check whether the buff is up or not.

File: engine/class_modules/sc_death_knight.cpp [hunk 4]
```diff
@@ -7710,15 +7710,16 @@ void death_knight_t::default_apl_frost()
   cooldowns -> add_action( "fireblood,if=buff.pillar_of_frost.remains<=8&buff.empower_rune_weapon.up" );
 
   // Pillar of Frost
-  cooldowns -> add_action( this, "Pillar of Frost", "if=cooldown.empower_rune_weapon.remains", "Frost cooldowns" );
+  cooldowns -> add_action( this, "Pillar of Frost", "if=cooldown.empower_rune_weapon.remains|!talent.breath_of_sindragosa.enabled", "Frost cooldowns" );
```
Comment (Melekus) [inline engine/class_modules/sc_death_knight.cpp]: If ERW is used on cooldown (cf. the L7717 addition), isn't it better to keep trusting that ERW is on cd before pressing PoF? It should be better to lose an ERW gcd to cast PoF than a PoF gcd to cast ERW, shouldn't it?

File: engine/class_modules/sc_death_knight.cpp (hunk 2 above)
Comment (nyterage) [reply engine/class_modules/sc_death_knight.cpp]: updated to reflect this change.

File: engine/class_modules/sc_death_knight.cpp (hunk 3 above)
Comment (nyterage) [reply engine/class_modules/sc_death_knight.cpp]: yeah, both have basically the same interaction. was just a rewrite of the previous lines in a cleaner, easier to read fashion, in my opinion anyways, as well as changing the time conditions from <8 to <=gcd. and yes, looks like checking for Unholy Strength being up should have been in there, i must have missed that interaction somehow. 
https://www.raidbots.com/simbot/report/6EUPF67Jf3oT87sgYs1hSS

File: engine/class_modules/sc_death_knight.cpp (hunk 4 above)
Comment (nyterage) [reply engine/class_modules/sc_death_knight.cpp]: While id be inclined to agree, the sims tend to disagree, not to much of an extent, but there is a difference. 
https://www.raidbots.com/simbot/report/ifDNmF9sGzZDrRQtsHpToQ

**Comment by nyterage** [comment]
Updated Sims, separating PTR and non PTR.

**PTR**

Best Icecap Talents
https://www.raidbots.com/simbot/report/mkYk82ct8pTsQAasiKDegt

FWF
https://www.raidbots.com/simbot/report/dBs4u15BforqPsmGsNknK9

**Non-PTR**

Best Icecap Talents
https://www.raidbots.com/simbot/report/fW7aetwz6rgSWMwKvCRmza

FWF
https://www.raidbots.com/simbot/report/2e65AFEzCUxf85MGwXcLjW

**Comment by nyterage** [comment]
Updated sims with the T25 Profile
Icecap (new default profile)
https://www.raidbots.com/simbot/report/fx3TBUJmKfMGt6oj3byeH1

Breath of Sindragosa T25 profile, Increase due to new Worldvein line. 
https://www.raidbots.com/simbot/report/2rTz94M8TAd68VESbkhSgh

**Comment by nyterage** [comment]
tons of sims to ensure no change/upgrades across everything changed. 

BoS, Worldvein major, FWF
https://www.raidbots.com/simbot/report/97NK9MJQFSvrL3Ej9s6D3L

BoS, BotE major, FWF
https://www.raidbots.com/simbot/report/rX4XTtAMcFNJb3gGNLgr3J

BoS, CLF major, FWF
https://www.raidbots.com/simbot/report/5rTE2qqFEiuZ95wg9kvWwV

BoS, BotE major, Gathering Storm
https://www.raidbots.com/simbot/report/4rAv81qDbmNadhmqDNX3rM

BoS Worldvein major, Gathering Storm
https://www.raidbots.com/simbot/report/u9LQfMCt1xHLUCAXH3Anso

BoS, CLF major, Gathering Storm
https://www.raidbots.com/simbot/report/3phUYV5SNkJ58onSCsqTQV

Icecap, Worldvein major, FWF
https://www.raidbots.com/simbot/report/au3PpKbCg5FeKFFYMvDZyG

Icecap, BotE major, FWF
https://www.raidbots.com/simbot/report/23ai7cJti9M8X1itMFdcJi

Icecap, CLF major, FWF
https://www.raidbots.com/simbot/report/qQRXccPuA3Hy2mLVVm3bcE

Icecap, Worldvein Major, Gathering Storm
https://www.raidbots.com/simbot/report/bytfKGsqXAdYzmYmpS9dAs

Icecap, BotE major, Gathering Storm
https://www.raidbots.com/simbot/report/c7qc3yBJyHW4CMgwdSiBrG

Icecap, CLF major, Gathering Storm
https://www.raidbots.com/simbot/report/7dV28PfBb3fW1njr6GZqXt

**Comment by Melekus** [comment]
Fixed the merge conflict

**Pull request merged**